lib/web_ui/lib/src/engine/keyboard_binding.dart

```diff
@@ -0,0 +1,479 @@
+
+// ASCII for a, z, A, and Z
+const int _kCharLowerA = 0x61;
+const int _kCharLowerZ = 0x7a;
+const int _kCharUpperA = 0x41;
+const int _kCharUpperZ = 0x5a;
+bool isAlphabet(int charCode) {
+  return (charCode >= _kCharLowerA && charCode <= _kCharLowerZ)
+      || (charCode >= _kCharUpperA && charCode <= _kCharUpperZ);
+}
+
+const String _kPhysicalCapsLock = 'CapsLock';
+
+const String _kLogicalDead = 'Dead';
+
+const int _kWebKeyIdPlane = 0x00800000000;
+const int _kAutogeneratedMask = 0x10000000000;
+
+// Bits in a Flutter logical event to generate the logical key for dead keys.
+//
+// Logical keys for dead keys are generated by annotating physical keys with
+// modifiers (see `_getLogicalCode`).
+const int _kDeadKeyCtrl = 0x100000000000;
+const int _kDeadKeyShift = 0x200000000000;
+const int _kDeadKeyAlt = 0x400000000000;
+const int _kDeadKeyMeta = 0x800000000000;
+
+typedef DispatchKeyData = bool Function(ui.KeyData data);
+
+/// Converts a floating number timestamp (in milliseconds) to a [Duration] by
+/// splitting it into two integer components: milliseconds + microseconds.
+Duration _eventTimeStampToDuration(num milliseconds) {
+  final int ms = milliseconds.toInt();
+  final int micro = ((milliseconds - ms) * Duration.microsecondsPerMillisecond).toInt();
+  return Duration(milliseconds: ms, microseconds: micro);
+}
+
+class KeyboardBinding {
+  /// The singleton instance of this object.
+  static KeyboardBinding? get instance => _instance;
+  static KeyboardBinding? _instance;
+
+  static void initInstance(html.Element glassPaneElement) {
+    if (_instance == null) {
+      _instance = KeyboardBinding._(glassPaneElement);
+      assert(() {
+        registerHotRestartListener(_instance!._reset);
+        return true;
+      }());
+    }
+  }
+
+  KeyboardBinding._(this.glassPaneElement) {
+    _setup();
+  }
+
+  final html.Element glassPaneElement;
+  late KeyboardConverter _converter;
+  final Map<String, html.EventListener> _listeners = <String, html.EventListener>{};
+
+  void _addEventListener(String eventName, html.EventListener handler) {
+    final html.EventListener loggedHandler = (html.Event event) {
+      if (_debugLogKeyEvents) {
+        print(event.type);
+      }
+      if (EngineSemanticsOwner.instance.receiveGlobalEvent(event)) {
+        return handler(event);
+      }
+    };
+    assert(!_listeners.containsKey(eventName));
+    _listeners[eventName] = loggedHandler;
+    html.window.addEventListener(eventName, loggedHandler, true);
+  }
+
+  /// Remove all active event listeners.
+  void _clearListeners() {
+    _listeners.forEach((String eventName, html.EventListener listener) {
+      html.window.removeEventListener(eventName, listener, true);
+    });
+    _listeners.clear();
+  }
+  bool _onKeyData(ui.KeyData data) {
+    bool? result;
+    EnginePlatformDispatcher.instance.invokeOnKeyData(data,
+      (bool handled) { result = handled; });
```

Comment:
Is this callback called synchronously? I'm worried that `_onKeyData` may return before the callback gets a chance to update `result`.

Reply:
It is supposed to. If the code is returned before the handler is called, `result` will be null and the code will throw on the `!` of the return statement. Let me add a comment to explain it.